src/newTest/java/org/opensearch/test/framework/cluster/SocketUtils.java

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright 2002-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.test.framework.cluster;
+
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.util.Random;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import javax.net.ServerSocketFactory;
+
+/**
+ * Simple utility methods for working with network sockets &mdash; for example,
+ * for finding available ports on {@code localhost}.
+ *
+ * <p>Within this class, a TCP port refers to a port for a {@link ServerSocket};
+ * whereas, a UDP port refers to a port for a {@link DatagramSocket}.
+ *
+ * @author Sam Brannen
+ * @author Ben Hale
+ * @author Arjen Poutsma
+ * @author Gunnar Hillert
+ * @author Gary Russell
+ * @since 4.0
+ */
+public class SocketUtils {
```

Comment:
Can we include the tests for this class?

Reply:
This class is used to find available port numbers. But the class makes no guarantees that the port is still available when the socket is bound to the port provided by the class. Therefore, I am afraid that the tests created for the class will be unstable. @peternied  do you think that writing such tests is necessary?

Reply:
I'd recommend mocking out the backend system so the behavior would be deterministic

Reply:
We will mock the backend systems  - still, we need to allocate ports for these in integtration tests.

Reply:
Writing unit tests even when integration tests are executed on code is useful to prevent non-obvious failures from creeping into the integration test runs.  I consider SocketUtils to have enough low-level complexity where this could be the case making the investment worthwhile, what do you think?

Reply:
Tests for SocketUtils have been added

Reply:
This class appears to be a copy of the same class in springframework: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/SocketUtils.html

From the docs it mentions:

> SocketUtils was introduced in Spring Framework 4.0, primarily to assist in writing integration tests which start an external server on an available random port. However, these utilities make no guarantee about the subsequent availability of a given port and are therefore unreliable. Instead of using SocketUtils to find an available local port for a server, it is recommended that you rely on a server's ability to start on a random port that it selects or is assigned by the operating system. To interact with that server, you should query the server for the port it is currently using.

Looks like springframework recommends the server start on a random port and the integration test query the server for the port. Should we consider springframework's recommendation here?